stdlib/Logging/src/ConsoleLogger.jl

```diff
@@ -116,8 +116,16 @@ function handle_message(logger::ConsoleLogger, level::LogLevel, message, _module
     end
 
     # Generate a text representation of the message and all key value pairs,
-    # split into lines.
-    msglines = [(indent=0, msg=l) for l in split(chomp(convert(String, string(message))::String), '\n')]
+    # split into lines.  This is specialised to improve type inferesence,
+    # and reduce the risk of resulting method invalidations.
+    msglines = if message isa Base.AnnotatedString
+        message = Base.AnnotatedString(String(message), Base.annotations(message))
```

Comment:
Seems like this is a place for using traits, as e.g. `applicable` and `!isempty`:
```suggestion
    message = string(message)
    annotations = applicable(Base.annotations, message) ? Base.annotations(message) : Base.annotations(Base.AnnotatedString(""))
    msglines = if !isempty(annotations)
        message = Base.AnnotatedString(convert(String, message)::String, annotations)
```

Reply:
Is it? I suppose this is predicated on the idea that there may be reimplementations of `AnnotatedString` that also implement `annotations` and want to be printed differently. I'm not sure how much thought we want to give to that?

Reply:
Maybe not too much. But I also gave some thought here to make a non-annotated AnnotatedString also just return a String to possibly simplify things later and try to DRY code slightly here. Nothing too critical though

Reply:
Perhaps calling `message = string(message)` before doing the `if message isa Base.AnnotatedString` might be a good balance of concerns?

Reply:
@vtjnash any further thoughts here?

Reply:
Yes, seems at least DRY that section of code would be good since if would fix some issues with that not branching correctly for things that render to an annotated string, even if you still use `isa(message, Base.AnnotatedString)` rather than changing it to `applicable(Base.annotations, message)` there.

The `isa` approach seems to have an issue already though that `SubString{AnnotatedString}` won't be handled correctly? I don't entirely know how annotation on a substring are currently implemented however

Reply:
Ah yes, I forgot about `SubString`s here ... which is something that's happened and been addressed before (via `_isannotated`). How do you feel about the change I just pushed?